src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -910,12 +910,27 @@ private static async Task<DurableOrchestrationStatus> GetDurableOrchestrationSta
                         }
                     }
 
-                    var counter = 0;
-                    indexList.Sort();
-                    foreach (var indexValue in indexList)
+                    if (eventTypes != null)
                     {
-                        historyArray.RemoveAt(indexValue - counter);
-                        counter++;
+                        for (int i = historyArray.Count - 1; i >= 0; i--)
+                        {
+                            var eventTypeName = historyArray[i].Value<string>("EventType");
+                            var eventType = (EventType)Enum.Parse(typeof(EventType), eventTypeName, true);
+                            if (!eventTypes.Contains(eventType))
+                            {
+                                historyArray.Remove(historyArray[i]);
+                            }
+                        }
+                    }
```

Comment:
Just to be sure I follow: why are we iterating the array backwards? Is it for performance reasons?

Reply:
Whatever the case, I think it would be good to document this in a comment :-)

Reply:
Thanks, David! The historyArray includes all kind of eventTypes. And here we only want to only leave the ones that callers ask for. So, this for loop is for this edit operation: to remove the ones that callers don't need. Since the remove and traverse operation happens at the same time, I iterate the array backwards so that the item number won't be changed due to these operations. :)

Reply:
That makes sense, thank you! Feel free to disregard my comment here then.